src/common/attribute.rs

```diff
@@ -0,0 +1,626 @@
+			x11_size += save_under.x11_size();
+			mask |= AttributeMask::SAVE_UNDER;
+		}
+
+		if let Some(event_mask) = &event_mask {
+			x11_size += event_mask.x11_size();
+			mask |= AttributeMask::EVENT_MASK;
+		}
+		if let Some(do_not_propagate_mask) = &do_not_propagate_mask {
+			x11_size += do_not_propagate_mask.x11_size();
+			mask |= AttributeMask::DO_NOT_PROPAGATE_MASK;
+		}
+
+		if let Some(colormap) = &colormap {
+			x11_size += colormap.x11_size();
+			mask |= AttributeMask::COLORMAP;
+		}
+
+		if let Some(cursor_appearance) = &cursor_appearance {
+			x11_size += cursor_appearance.x11_size();
+			mask |= AttributeMask::CURSOR_APPEARANCE;
+		}
+
+		Self {
+			x11_size,
+			mask,
+
+			background_pixmap,
+			background_pixel,
+
+			border_pixmap,
+			border_pixel,
+
+			bit_gravity,
+			win_gravity,
+
+			backing_store,
+			backing_planes,
+			backing_pixel,
+
+			// These booleans are converted into our [`Bool`] type so that they can easily be
+			// written as four bytes.
+			override_redirect: override_redirect.map(std::convert::Into::into),
+			save_under: save_under.map(Into::into),
+
+			event_mask,
+			do_not_propagate_mask,
+
+			colormap,
+
+			cursor_appearance,
+		}
+	}
+
+	#[must_use]
+	pub const fn background_pixmap(&self) -> &Option<ParentRelatable<Option<Pixmap>>> {
+		&self.background_pixmap
+	}
+	#[must_use]
+	pub const fn background_pixel(&self) -> &Option<Pixel> {
+		&self.background_pixel
+	}
+
+	#[must_use]
+	pub const fn border_pixmap(&self) -> &Option<CopyableFromParent<Pixmap>> {
+		&self.border_pixmap
+	}
+	#[must_use]
+	pub const fn border_pixel(&self) -> &Option<Pixel> {
+		&self.border_pixel
+	}
+
+	#[must_use]
+	pub const fn bit_gravity(&self) -> &Option<BitGravity> {
+		&self.bit_gravity
+	}
+	#[must_use]
+	pub const fn win_gravity(&self) -> &Option<WinGravity> {
+		&self.win_gravity
+	}
+
+	#[must_use]
+	pub const fn backing_store(&self) -> &Option<BackingStore> {
+		&self.backing_store
+	}
+	#[must_use]
+	pub const fn backing_planes(&self) -> &Option<u32> {
+		&self.backing_planes
+	}
+	#[must_use]
+	pub const fn backing_pixel(&self) -> &Option<Pixel> {
+		&self.backing_pixel
+	}
+
+	#[must_use]
+	pub fn override_redirect(&self) -> Option<bool> {
+		self.override_redirect.map(Into::into)
+	}
+	#[must_use]
+	pub fn save_under(&self) -> Option<bool> {
+		self.save_under.map(Into::into)
+	}
+
+	#[must_use]
+	pub const fn event_mask(&self) -> &Option<EventMask> {
+		&self.event_mask
+	}
+	#[must_use]
+	pub const fn do_not_propagate_mask(&self) -> &Option<DeviceEventMask> {
+		&self.do_not_propagate_mask
+	}
+
+	#[must_use]
+	pub const fn colormap(&self) -> &Option<CopyableFromParent<Colormap>> {
+		&self.colormap
+	}
+
+	#[must_use]
+	pub const fn cursor_appearance(&self) -> &Option<Option<CursorAppearance>> {
+		&self.cursor_appearance
+	}
+}
+
+impl X11Size for Attributes {
+	fn x11_size(&self) -> usize {
+		self.x11_size
+	}
+}
+
+impl Readable for Attributes {
+	#[allow(clippy::too_many_lines)]
```

Comment:
Due to the repetitive nature of the reads:
```rust
let $attribute_name = if mask.contains(AttributeMask::$ATTRIBUTE_NAME) {
    let ret = <$AttributeType>::read_from(buf)?;
    x11_size += ret.x11_size();

    Some(ret)
} else {
    None
}
```
(declarative macro syntax there just for demonstration), perhaps the lint is correct, and the function can be made shorter by defining a closure to do this.

```rust
let try_read_attr = |attr_mask| if mask.contains(attr_mask) {
    let ret = /* err... can closures use generics? */::read_from(buf)?;
    x11_size += ret.x11_size();

    Some(ret)
} else {
    None
};
```